Add option to write results into a single zip archive

flomnes opened the pull request:
This features adds the possibility to write results to a single .zip file. There was a lot of refactoring to centralize output, since before that everything was based on files & directories. I created an interface class & a factory method, depending on a new parameter.

### TODO
- [x] Move minizip-ng to antares-deps #814
- [x] MPS files, and other Xpansion-related files #807
- [x] Annual statistics (checkIntegrity.txt, etc.) #808
- [x] Export about-the-study #804 
- [x] simulation.log #805
- [x] hydro debug files #809
- [x] Export generated timeseries to the .zip archive if "store in output" is enabled
- [x] Export generated timeseries when the `-g` option is provided
- [x] Avoid copies when possible, switch from (buffer, size) to `std::string` ?
- [x] Fix failing linking in the Sonar job #810
- [x] Performance tests : check if we actually save time by saving the results into a single zip archive.
- [x] Document the feature (command-line option & study parameter) #813
- [x] Document the [study format change](https://github.com/AntaresSimulatorTeam/Antares_Simulator/blob/develop/docs/reference-guide/13-file-format.md) #813
- [x] Handle errors from minizip-ng (not enough space on disk, incorrect path, etc.) #822
- [ ] Track and fix bugs
- [x] Make sure that a valid .zip file is produced the program is terminated (Ctrl+C, etc.), see [SO](https://stackoverflow.com/questions/2350489/how-to-catch-segmentation-fault-in-linux) #827
- [x] Write an Architecture Decision Record (ADR)
- [ ] Add unit tests
- [X] TS numbers
- [ ] Handle case where archive / directory already exists

flomnes committed:
Save output to .zip archive

flomnes committed:
[CI] Try to fix missing libzip

flomnes committed:
Write most output files into zip archive

This is going to break tests very hard

flomnes committed:
Fix path for mc-ind/mc-all

flomnes committed:
Add MPS files to archive

flomnes committed:
Pass archive pointer to ISimulation

flomnes committed:
Add class ZipWriteJob

flomnes committed:
Fix warning related to order of data members

flomnes committed:
Add zip writer jobs

flomnes committed:
Use minizip-ng instead of libzippp

flomnes committed:
Adjust compression level & method

flomnes committed:
Set modified date properly

flomnes committed:
Fix .zip path

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/thread-writer-zip

flomnes committed:
Some renaming, add const qualifier

flomnes committed:
Some more renaming

flomnes committed:
Add interface class for result writer

flomnes committed:
Restart queue for mc-all, write thermal.txt, etc.

flomnes committed:
Add immediate write info files, factory, and parameters I/O

flomnes committed:
Rename headers, clean code, CMakeFiles and GH workflows

flomnes committed:
Some more cleaning

flomnes committed:
Force zip64

flomnes committed:
Add extern "C" around C header files

flomnes committed:
Avoid copies by using std::move

flomnes committed:
Add option -zip-output / -z

flomnes committed:
Disable unused minizip deps

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/thread-writer-zip

flomnes committed:
Add zip-specific benchmarking info (write/wait)

flomnes committed:
Write generated thermal TS into .zip when "store in output" is ON

TODO : hydro / wind / solar / load

flomnes committed:
Add simple version for Matrix<T>::saveToBuffer

flomnes committed:
Write generated hydro TS into .zip when "store in output" is ON

TODO : wind / solar / load

flomnes committed:
Remove useless namespaces

flomnes committed:
Write generated win /solar/load TS into .zip when "store in output"

flomnes committed:
Fix empty .zip when "gen TS only" option (-g) is used

flomnes committed:
Write execution_info.ini in the .zip file

flomnes committed:
Remove Benchmarking::FileWriter and Benchmarking::iniFilewriter

use conversion function instead

flomnes committed:
Document zip output feature

flomnes committed:
Use antares-deps to provide minizip-ng

flomnes committed:
Fix previous commit

flomnes committed:
Bump antares-deps 2.0.3-rc1 -> 2.0.3-rc2

flomnes committed:
Empty commit

flomnes committed:
Merge pull request #814 from AntaresSimulatorTeam/feature/minizip-use-deps

Use antares-deps to provide minizip-ng

flomnes committed:
Bump antares-deps 2.0.3-rc2 -> 2.0.3

guilpier-code commented on src/libs/antares/study/fwd.h:
```diff
@@ -540,6 +540,14 @@ const char* DayAheadReserveManagementModeToCString(DayAheadReserveManagement daR
+enum ResultFormat
+{
+    // Store outputs as files inside directories
+    legacyFilesDirectories = 0,
```
I see what you mean with **legacy**, but someone else does not necessarily know. It is supposed to mean text files.
**legacyFilesDirectories** ==> **textFiles** ?

guilpier-code commented on src/libs/antares/study/parameters.h:
```diff
@@ -557,6 +557,8 @@ class Parameters final
     OrtoolsSolver ortoolsEnumUsed;
     //@}
+    // Format of results. Currently, only single files or zip archive are supported
+    ResultFormat resultFormat = legacyFilesDirectories;
```
There are often many result**s** (with an **s**), even if that code is already understandable

guilpier-code commented on src/libs/antares/study/study.cpp:
```diff
@@ -774,22 +775,23 @@ bool Study::prepareOutput()
-        logs.error() << "I/O: impossible to create the directory " << buffer;
-        return false;
-    }
+    IO::Directory::Create(buffer);
```
Why did you get rid of : 
```C++
if (not IO::Directory::Create(buffer))
{
        logs.error() << "I/O: impossible to create the directory " << buffer;
        return false;
 }
```
?

flomnes replied:
Because directories are created automatically

guilpier-code commented on src/solver/writer/i_writer.h:
```diff
@@ -0,0 +1,23 @@
+{
+public:
+    using Ptr = std::shared_ptr<IResultWriter>;
+    virtual void addJob(const std::string& entryPath, Yuni::Clob& entryContent) = 0;
```
Functions **addJob** actually flushes the results on disk, into a file format or another (.txt or .zip). Why calling them **addJob** rather than **flush** or **save** or **dump** ?

guilpier-code replied:
Maybe you checked it, but I does not seem that we need **addJob** in **Yuni::Clob** version, or we can avoid it. I did not make a full check...
In case I'm right, we would get rid of a method and get rid of the **Yuni** dependance

flomnes replied:
Sadly, it's needed.

flomnes replied:
> Functions **addJob** actually flushes the results on disk, into a file format or another (.txt or .zip). Why calling them **addJob** rather than **flush** or **save** or **dump** ?

Because the file content isn't saved immediately, or flushed.

It's only written immediately if the "files" version is used.

guilpier-code replied:
ok

guilpier-code commented on src/solver/simulation/solver.hxx:
```diff
@@ -265,7 +264,10 @@ inline ISimulation<Impl>::ISimulation(Data::Study& study,
+ pDurationCollector(duration_collector),
+ qs(std::make_shared<Yuni::Job::QueueService>()),
+ pResultWriter(
+   resultWriterFactory(study.parameters.resultFormat, study.folderOutput, qs, duration_collector))
```
**qs** ==> **qs_**, **_qs** or **pQs**

guilpier-code replied:
Furthermore, why making **qs** a smart pointer ? Why not a reference data member ?

flomnes replied:
To avoid dangling references

guilpier-code replied:
Ok

guilpier-code commented on src/solver/writer/zip_writer.cpp:
```diff
@@ -0,0 +1,106 @@
+    EnsureQueueStartedIfNeeded ensureQueue(this, pQueueService);
+    std::string buffer;
+    entryContent.saveToString(buffer);
+    pQueueService->add(new ZipWriteJob<std::string>(*this, entryPath, buffer, pDurationCollector),
```
Can't we replace ini files **addJob**'s body with something like : 
```C++
void ZipWriter::addJob(const std::string& entryPath, Antares::IniFile& entryContent)
{
    std::string buffer;
    entryContent.saveToString(buffer);
    addJob(entryPath, buffer);
}
```
and this to avoid some kind of code duplication ?

flomnes replied:
Done, in another PR. Thanks

guilpier-code commented on src/solver/writer/zip_writer.cpp:
```diff
@@ -0,0 +1,106 @@
+{
+    mz_zip_file file_info;
+
+    Benchmarking::Timer timer_wait;
```
Here we have 3 kinds of tasks got done : time measurement and compress job.
We could separate them like this (it's an example) : 
```C++
template<class ContentT>
void ZipWriteJob<ContentT>::onExecute()
{
    // ok, I don't see how to split this block
    Benchmarking::Timer timer_wait;
    std::lock_guard<std::mutex> guard(pZipMutex); // Wait
    timer_wait.stop();
    pDurationCollector->addDuration("zip_wait", timer_wait.get_duration());

    Benchmarking::Timer timer_write;
    mz_zip_file file_info = prepareFileInfo();  // I guess preparing file is part of compressing, so it's measured as well
    zip_content(file_info);
    timer_write.stop();
    pDurationCollector->addDuration("zip_write", timer_write.get_duration());
}
```
where **prepareFileInfo()** and **zip_content(...)** would be private method of class **ZipWriteJob**

guilpier-code commented on src/solver/writer/zip_writer.h:
```diff
@@ -0,0 +1,68 @@
+              const char* archivePath,
+              Benchmarking::IDurationCollector* duration_collector);
+    ~ZipWriter();
+    void addJob(const std::string& entryPath, Yuni::Clob& entryContent) override;
+    void addJob(const std::string& entryPath, std::string& entryContent) override;
+    void addJob(const std::string& entryPath, Antares::IniFile& entryContent) override;
```
So **addJob** methods are made to run jobs. 
Calling them **addJob** could be confusing, it gives the idea that we add some job to the queue and that we run the jobs later. 
Why not call them **run(...)** or **write(...)** ?

guilpier-code replied:
But a question bothers me : 
Why do we use jobs to do the work ? Jobs are launched in a different thread than the calling thread. Why don't we execute the work of adding the zip archive a piece of data in the same thread than the calling thread ?
And why don't we run a new job for writing classic txt files ?

flomnes replied:
> Why don't we execute the work of adding the zip archive a piece of data in the same thread than the calling thread ?

Because zip files can't be written concurrently, _ie_ since the lib isn't thread-safe. All operations on it are `mutex`ed. If we wrote results immediately in this case, a lot of CPU time would be wasted waiting for write operations to be completed. We don't want that, so we give the responsability of writing data to a separate thread.

guilpier-code commented on src/libs/antares/benchmarking/file_content.h:
```diff
@@ -19,13 +19,16 @@ class FileContent
     void addDurationItem(const std::string& name, unsigned int duration, int nbCalls);
 
+    template<class DestT>
+    DestT to();
```
What does this function do exactly (I mean : **to()**) ?
After having looked, I see that it converts the content into an ini file content.
- The name is unclear to me, we should rename this function (**convertIntoFormat\<TargetFormatType\>()**)
- Why not dropping the template parameter and call the function **convertIntoIniContent()** 
- By adding this function, we make the class grow, it starts doing something it shouldn't do, that is converting the content into another kind of content (ini content). We should start to think of extracting the conversion into another function or class, for example by introducing a **decorator** that would make the conversions. If some other conversion is made, we would not have to touch the **FileContent** class.

guilpier-code commented on src/solver/writer/zip_writer.h:
```diff
@@ -0,0 +1,68 @@
+namespace Solver
+{
+class ZipWriter;
+template<class ContentT>
```
**ContentT** ==> **ContentType**

guilpier-code commented on src/solver/simulation/solver.hxx:
```diff
@@ -265,7 +264,10 @@ inline ISimulation<Impl>::ISimulation(Data::Study& study,
- pDurationCollector(duration_collector)
+ pDurationCollector(duration_collector),
+ qs(std::make_shared<Yuni::Job::QueueService>()),
+ pResultWriter(
```
I would have put **pResultWriter** initialization in the constructor body to highlight it a bit more, and because it is a more complex initialization than just affecting a value, but it's a question of taste I suppose.

guilpier-code replied:
Reminder : I said somewhere else that, **pResultWriter** is shared with (at least) **Application** class, it should be imported from outside the **ISimulation** constructor

guilpier-code commented on src/solver/main/economy.cpp:
```diff
@@ -51,6 +51,7 @@ void Application::runSimulationInEconomicMode()
 
         this->pOptimizationInfo = simulation.getOptimizationInfo();
     }
+    this->pResultWriter = simulation.pResultWriter;
```
I think we should not make this kind of initialization (line 54), as if it was done behind the scene.
In this case, **pResultWriter** should be initialized in a method class **Application**, as early as possible in the execution, and then passed to the simulation : 
```C++
   // See line 42
   SimulationType simulation(*pStudy, pSettings, &pDurationCollector, pResultWriter);
```
Same remark for mode **adequacy** and **draft**.

flomnes replied:
I agree. But since the output folder isn't known when `Application` is created, we'd have to provide an `initialize` method, instead if providing it to the constructor.

guilpier-code replied:
ok. As I say in PR #804, the study itself needs the writer to write the **about-the-study** things.
So I guess we should initialize the writer somewhere in the study as soon as the kind of writer is known.

guilpier-code commented on src/solver/writer/immediate_file_writer.h:
```diff
@@ -0,0 +1,26 @@
+#pragma once
```
Why calling the txt result writer the immediate result writer ? Why **immediate** ?
Can't we just call it **txtResultsWriter** ?

guilpier-code replied:
Same remark for the source file name

guilpier-code commented on src/solver/writer/immediate_file_writer.cpp:
```diff
@@ -0,0 +1,78 @@
+{
+    auto output = prepareDirectoryHierarchy(pOutputFolder, entryPath);
+    std::string buffer;
+    entryContent.saveToString(buffer);
```
Regarding : 
```C++
std::string buffer;
entryContent.saveToString(buffer);
```
We could expect : 
```C++
std::string buffer = entryContent.intoString();
``` 
I guess a call to std::string move consrtuctor will be made when returning a std::string in **intoString()**

guilpier-code commented on src/libs/antares/study/mps.cpp:
```diff
@@ -37,9 +37,9 @@ namespace Antares
-FILE* Study::createFileIntoOutputWithExtension(const YString& prefix,
-                                               const YString& extension,
-                                               uint numSpace) const
+std::string Study::createFileIntoOutputWithExtension(const YString& prefix,
```
We should rename this function to stick to what it really does, for instance into **createOutputFileName(...)**

guilpier-code commented on src/solver/application.cpp:
```diff
@@ -551,6 +556,9 @@ void Application::writeExectutionInfo()
     pTotalTimer.stop();
     pDurationCollector.addDuration("total", pTotalTimer.get_duration());
 
+    if (!pResultWriter)
```
We could have : 
```C++
pTotalTimer.stop(); // If we have a embarrassing return, we should stop first the timer, right ?

if (!pStudy || !pResultWriter)
    return;

pDurationCollector.addDuration("total", pTotalTimer.get_duration());
```

guilpier-code commented on src/solver/ts-generator/xcast/xcast.cpp:
```diff
@@ -85,21 +89,18 @@ void XCast::exportTimeSeriesToTheOutput(Progression::Task& progression, Predicat
         String filename;
```
I don't know why the current function **exportTimeSeriesToTheOutput**(...)  in a member of class **XCast** : it does not use a data / function member of  **XCast** (apart from **study**, ..., and now **pWriter**, ok).
Maybe we should make it an external function and so make **XCast** thinner. Ok, not a priority.

guilpier-code commented on src/solver/ts-generator/hydro.cpp:
```diff
@@ -311,19 +314,23 @@ bool GenerateHydroTimeSeries(Data::Study& study, uint currentYear)
+                    writer->addJob(output.c_str(), storage_buffer);
                 }
 
+                area.hydro.series->flush();
```
As far I can see, **flush()** does nothing (empty body), let's remove it

guilpier-code commented on src/solver/ts-generator/hydro.cpp:
```diff
@@ -311,19 +314,23 @@ bool GenerateHydroTimeSeries(Data::Study& study, uint currentYear)
+                    std::string ror_buffer;
                     output.clear() << study.buffer << SEP << "ror.txt";
-                    area.hydro.series->ror.saveToCSVFile(output);
+                    writer->addJob(output.c_str(), ror_buffer);
```
I don't understand how the following instruction can work : 
```c++
writer->addJob(output.c_str(), ror_buffer);
```
`area.hydro.series->ror` should be turned into a buffer, and then passed to **addJob**

guilpier-code commented on src/solver/ts-generator/hydro.cpp:
```diff
@@ -311,19 +314,23 @@ bool GenerateHydroTimeSeries(Data::Study& study, uint currentYear)
-                    area.hydro.series->storage.saveToCSVFile(output);
-
-                    area.hydro.series->flush();
+                    writer->addJob(output.c_str(), storage_buffer);
```
Same as above

guilpier-code commented on src/solver/ts-generator/thermal.cpp:
```diff
@@ -157,42 +166,22 @@ void GeneratorTempData::writeResultsToDisk(const Data::Area& area,
-            folder << study.folderOutput << SEP << "ts-generator" << SEP << "thermal" << SEP
-                   << "mc-" << year << SEP << area.id;
+        pWriter->addJob(pTempFilename.c_str(), buffer);
+    }
```
Please remove **prepareOutputFoldersForAllAreas** declaration in class **GeneratorTempData**

guilpier-code commented on src/solver/utils/mps_utils.cpp:
```diff
@@ -16,14 +17,14 @@ using namespace Antares::Data;
 constexpr size_t OPT_APPEL_SOLVEUR_BUFFER_SIZE = 256;
 
-static void printHeader(FILE* Flot, int NombreDeVariables, int NombreDeContraintes)
+static void printHeader(Clob& Flot, int NombreDeVariables, int NombreDeContraintes)
```
In this source file, we could replace **Clob** with **std::string**, we would get rid of Yuni and the code would be faster to execute (no attention to performance issues was given in **Yuni::Clob**)

flomnes replied:
The problem is that as of C++17, `std::string` provides no formatting capabilities, for example `"%4.4f"`

guilpier-code commented on src/solver/variable/surveyresults/surveyresults.cpp:
```diff
@@ -139,7 +139,7 @@ void InternalExportDigestLinksMatrix(const Data::Study& study,
-void ExportGridInfosAreas(const Data::Study& study, const String& folder)
+static void ExportGridInfosAreas(const Data::Study& study, IResultWriter::Ptr writer)
 {
     Clob out;
```
We should try to remove **Yuni::Clob** from this function, and from this file as well (if it's easy to do and test), and replace it with **std::string**, faster at execution

flomnes replied:
Sure

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/thread-writer-zip

flomnes committed:
Fix build

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/thread-writer-zip

flomnes committed:
Rename qs -> pQueueService

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/thread-writer-zip

flomnes committed:
I/O error handling (#822)

* Handle I/O errors in zip writer

* Immediate write : handle I/O errors

* Update src/solver/writer/immediate_file_writer.cpp

* Update src/solver/writer/immediate_file_writer.cpp

* [skip ci] Format & simplify code

* Fix build

* PR remarks

flomnes committed:
Windows CI : fix overlapping CMake dependencies (zlib) (#865)

* Fix overlapping zlib, 1st try

* Revert "Fix overlapping zlib, 1st try"

This reverts commit ceec281cc70e93cf7b39d8f85d9046ac884e3171.

* Use antares-deps 2.0.4-rc1

* Use antares-deps 2.0.4-rc2

* Use antares-deps 2.0.4

* minizip -> minizip-ng

* Update antares-deps submodule

* Revert multiple commits

* Fix minizip

* Revert to antares-deps 2.0.2

* deps 2.0.3

* Fix MZ flags

* Force build of MZ in Windows

* Fix configure command for Win

* Use antares-deps 2.0.2

* Use antares-deps 2.0.2 (no minizip)

* Enable code coverage

flomnes committed:
PR remarks

flomnes committed:
[output/result-format] files => txt-files

flomnes committed:
Merge remote-tracking branch 'github/feature/thread-writer-zip' into doc/zip-output

flomnes committed:
13-file-format.md

flomnes committed:
Merge remote-tracking branch 'github/develop' into doc/zip-output

flomnes committed:
Add missing section

flomnes committed:
Merge pull request #813 from AntaresSimulatorTeam/doc/zip-output

Document zip output feature

flomnes committed:
Save about-the-study and files therein using generic result writer (#804)

* Save about-the-study and files using generic result writer

* Add getter for writer

* Use generic writer to write problem.mps, criterion, interco, etc.

Those files are needed by Xpansion

* Use generic writer for annual statistics (#808)

* Some renaming

* PR remarks

* Finalize write upon receiving SIGINT / SIGTERM (#827)

* Bump to CXX 17

* Clarify execution-info.ini file

* Fix missing renewable columns in districts (sets of areas) (#802)

* execution_info.ini : add OR-Tools used & solver (#820)

* execution_info.ini : add OR-Tools used & solver

* PR remarks

* Fix wrong number of cores + code de-duplication (#793)

* Fix wrong number of cores + code de-duplication

* PR remarks + small refactor (template)

* Release/8.3.1 (#818)

* Finalize write upon receiving SIGINT / SIGTERM

This should ensure that a valid zip archive is produced even if the
user pushes "Ctrl+C".

* Avoid double-free

* Some formatting

* Refactor + add logs

* Simplify tuneSolverSpecificOptions (#829)

Use MPSolver::ProblemType

* [skip ci] Remove useless {}

* Restore notifications from Sonar (#828)

Source: https://github.com/AntaresSimulatorTeam/antares-xpansion/pull/501

* Statistics: use std::mutex / std::atomic<> to prevent concurrent writes (#838)

* Statistics: use std::mutex / std::atomic<> to prevent concurrent writes

* Apply PR remarks + renaming

* Base for BC dual values (#749)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* doc : new input and output format due to addition of BC marginal prices (#836)

* doc : new input and output format due to addition of BC marginal prices

* BC marginal prices doc : the folder containing the binding constraints output is renamed

* BC marginal price (#801)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

* Revert "BC marginal price (#801)"

This reverts commit a6fffd91c0087db4cb3d4532989f3adbb9e44047.

* Feature/bc marginal price (#845)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

* BC. MARG. COST: use 2 decimals for the output value (#855)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

* BC. MARG. COST: use 2 decimals for the output value

Reading the changelog of version 6.5, I came accros this fix note in
version 6.5

> - Output: the link-variable "MARG.COST" was rounded to an integer
  value (changed to 2 decimal accuracy)

So I thought that we might use 2 decimals for BC. MARG. COST as well.

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

* Fix potential segfault with many BCs, refactor (#853)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

* Fix potential segfault with many BCs, refactor

* Add static_cast<uint>

* Remove lambda return type

* Smell code

* Add TODO note

* PR remarks

* Use std::distance

* Code smells

* More code smells

* Remove unused member

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

* Move Application::installSignalHandlers -> free function

* Add forgotten header file

* Use template helper function

* Fix build

Co-authored-by: guilpier-code <62292552+guilpier-code@users.noreply.github.com>
Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

* Remove result writer setter/getter from Study

* Fix link

* Init writer in study

* Fix overlapping zlib, 1st try

* Revert "Fix overlapping zlib, 1st try"

This reverts commit ceec281cc70e93cf7b39d8f85d9046ac884e3171.

* Use antares-deps 2.0.4-rc1

* Use antares-deps 2.0.4-rc2

* Add missing std:: (fix build)

* Add forward decl for Benchmarking::IDurationCollector

* Add & call Study::prepareWriter

* Format code, fix build

* Fix simple-study unit test

* Use antares-deps 2.0.4

* In-class init for pQueueService

* minizip -> minizip-ng

* Update antares-deps submodule

* Revert multiple commits

* Fix minizip

* Revert to antares-deps 2.0.2

* deps 2.0.3

* Fix MZ flags

* Force build of MZ in Windows

* Fix configure command for Win

* Use antares-deps 2.0.2

* Use antares-deps 2.0.2 (no minizip)

* Enable code coverage

* Add try..catch to handle no queue/no result writer

* Remove check (moved to init)

* Save "about the study" earlier

* Rename simulation -> simulationMetadata

* Use generic writer to write simulation.log (#805)

* Use generic writer to write simulation.log

* PR remarks

* MPS : don't attempt to create directory

* Fix conficts

* Use generic writer to write MPS files, criterion, interco, etc. (#807)

* Use generic writer for hydro debug files (#809)

* Use generic writer for hydro debug files

* PR remarks

* Remove result writer setter/getter from Study

* Fix link

* Init writer in study

* Add missing std:: (fix build)

* Add forward decl for Benchmarking::IDurationCollector

* Add & call Study::prepareWriter

* Format code, fix build

* Fix simple-study unit test

* In-class init for pQueueService

* PR remarks

* Add try..catch to handle no queue/no result writer

* Remove check (moved to init)

* Fix build

* Move writer creation to Study (#864)

* Remove result writer setter/getter from Study

* Fix link

* Init writer in study

* Add missing std:: (fix build)

* Add forward decl for Benchmarking::IDurationCollector

* Add & call Study::prepareWriter

* Format code, fix build

* Fix simple-study unit test

* In-class init for pQueueService

* Add try..catch to handle no queue/no result writer

* Remove check (moved to init)

* Remove unnecessary checks

* Fix build

* PR remarks

* Rename saveMiscFilesIntoOutput -> saveAboutTheStudy

Co-authored-by: guilpier-code <62292552+guilpier-code@users.noreply.github.com>
Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

flomnes committed:
Reorder init in Study::Study (comp. warning)

flomnes committed:
Rename addJob -> addEntry

flomnes committed:
Clean up zip_writer.cpp

flomnes committed:
Small refactor

flomnes committed:
Remove addEntry with ini file, conversion method

flomnes committed:
Do not create directories

flomnes committed:
Remove TODO note

flomnes committed:
Export digest.txt

flomnes committed:
Remove useless Antares::Solver::

flomnes committed:
Fix path for digest.txt

flomnes committed:
Remove unused methods

flomnes committed:
Fix path for areas.txt, links.txt and thermal.txt

flomnes committed:
Fix \t <-> \n

flomnes committed:
Rename addEntry -> addEntryFromBuffer

flomnes committed:
Add and use addEntryFromFile

flomnes committed:
Revert antares-deps

flomnes committed:
Fix write for about-the-study/map

flomnes committed:
Don't continue on failed tests (ubuntu)

flomnes committed:
Don't export hydro-debug by default, or if no writer exists

flomnes committed:
Fix non-passing tests

flomnes committed:
Simplify implementation of getFilenameWithExtension

flomnes committed:
Bump minizip-ng 3.0.6 -> 3.0.7, disable code coverage

flomnes committed:
Move MZ_CODE_COVERAGE to SonarCloud job

flomnes committed:
Apply clang-format to diff /w develop

git diff github/develop..@ --name-only | xargs ./src/format-code.sh

flomnes committed:
[TMP] Run tests sequentially

guilpier-code committed:
Fix : in result files' path, change a '/' in to a SEP, valid on all platforms (#926)

flomnes committed:
Revert "[TMP] Run tests sequentially"

This reverts commit d9ac5f470517b563d2acf5fb95f8bc89b4017d50.

flomnes committed:
Use generic writer to write ts-numbers, refactor (#929)

* Use generic writer to write ts-numbers, refactor

* Fix SonarQube issues

flomnes committed:
Merge remote-tracking branch 'github/develop'

flomnes committed:
Fix build

flomnes committed:
OR-Tools : remove temporary MPS files

flomnes committed:
Move deletion to dedicated static function

flomnes committed:
ADR for zip

flomnes committed:
Restore print dimensions for TS numbers
